timemachine/fe/absolute_hydration.py

```diff
@@ -155,3 +173,184 @@ def log_prob(xs, lam):
+        Forcefield to be used for the system
+
+    host_config: HostConfig
+        Configuration for the host system.
+
+    n_frames: int
+        number of samples to generate for each lambda window, where each sample is `steps_per_frame` steps of MD.
+
+    prefix: str
+        A prefix to append to figures
+
+    seed: int
+        Random seed to use for the simulations.
+
+    n_windows: None
+        Number of windows used for interpolating the the lambda schedule with additional windows.
+
+    keep_idxs: list of int or None
+        If None, return only the end-state frames. Otherwise if not None, use only for debugging, and this
+        will return the frames corresponding to the idxs of interest.
+
+    n_eq_steps: int
+        Number of equilibration steps for each window.
+
+    steps_per_frame: int
+        The number of steps to take before collecting a frame
+
+    Returns
+    -------
+    SimulationResult
+        Collected data from the simulation (see class for storage information). We currently return frames
+        from only the first and last window.
+    """
+    bt = BaseTopology(mol, ff)
+    afe = AbsoluteFreeEnergy(mol, bt)
+
+    # note: tm convention lambda=1 means "decoupled", lambda=0 means "coupled"
+    lambda_schedule = construct_pre_optimized_absolute_lambda_schedule_solvent(n_windows)[::-1]
+    assert np.isclose(lambda_schedule[0], 1.0) and np.isclose(lambda_schedule[-1], 0.0)
+
+    temperature = DEFAULT_TEMP
+    initial_states = setup_initial_states(afe, ff, host_config, temperature, lambda_schedule, seed)
+    protocol = SimulationProtocol(n_frames=n_frames, n_eq_steps=n_eq_steps, steps_per_frame=steps_per_frame)
+
+    if keep_idxs is None:
+        keep_idxs = [0, len(initial_states) - 1]  # keep first and last windows
+    assert len(keep_idxs) <= len(lambda_schedule)
+
+    combined_prefix = get_mol_name(mol) + "_" + prefix
+    try:
+        return estimate_free_energy_given_initial_states(
+            initial_states, protocol, temperature, combined_prefix, keep_idxs
+        )
+    except Exception as err:
+        with open(f"failed_ahfe_result_{combined_prefix}.pkl", "wb") as fh:
+            pickle.dump((initial_states, protocol, err), fh)
+        raise err
+
+
+def setup_initial_states(
+    afe: AbsoluteFreeEnergy,
+    ff: Forcefield,
+    host_config: HostConfig,
+    temperature: float,
+    lambda_schedule: Array,
+    seed: int,
+) -> List[InitialState]:
+    """
+    Setup the initial states for a series of lambda values. It is assumed that the lambda schedule
+    is a monotonically increasing sequence in the closed interval [0, 1].
+
+    Parameters
+    ----------
+    afe: AbsoluteFreeEnergy
+        An AbsoluteFreeEnergy object which contains the mol structure
+
+    host_config: HostConfig
+        Configurations of the host.
+
+    temperature: float
+        Temperature to run the simulation at.
+
+    lambda_schedule: list of float of length K
+        Lambda schedule.
```

Comment:
nit[style opinion] -- parameter descriptions of this form can be omitted.

Consider either removal or expansion. (E.g. "must start at 1 and end at 0")